src/rust/engine/fs/store/src/lib.rs

```diff
@@ -598,6 +597,48 @@ impl Store {
       .to_boxed()
   }
 
+  /// Load a REv2 Tree from a remote CAS and cache the embedded Directory protos in the
+  /// local store. Tree is used by the REv2 protocol as an optimization for encoding the
+  /// the Directory protos that compromose the output directories from a remote
+  /// execution reported by an ActionResult.
+  ///
+  /// Returns an Option<Digest> representing the `root` Directory within the Tree (if it
+  /// in fact exists in the remote CAS).
+  ///
+  /// This method requires that this Store be configured with a remote CAS (and will return
+  /// an error if this is not the case).
+  pub async fn load_tree_from_remote(&self, tree_digest: Digest) -> Result<Option<Digest>, String> {
+    let remote = if let Some(ref remote) = self.remote {
+      remote
+    } else {
+      return Err("Cannot load Trees from a remote without a remote".to_owned());
+    };
+
+    let tree_opt = remote
+      .load_bytes_with(EntryType::Directory, tree_digest, |b| {
+        let mut tree = Tree::new();
+        tree
+          .merge_from_bytes(&b)
+          .map_err(|e| format!("protobuf decode error: {:?}", e))?;
+        Ok(tree)
+      })
+      .await?;
+
+    let tree = match tree_opt {
+      Some(t) => t,
+      None => return Ok(None),
+    };
+
+    // Cache the returned `Directory` proto and the children `Directory` protos in
+    // the local store.
+    let root_digest = self.record_directory(tree.get_root(), true).await?;
```

Comment:
Ideally this would use `join_all` (or some async/await version of the combinator) so these can all be done in parallel, as they don't depend on each other :)

Reply:
Good point! Will update to do that.